### PR TITLE
Pagination for Nested Commenting & DBMS bug fixes 

### PR DESCRIPTION
**Pull request by thisaltennakoon**

### CODE REVIEW

**Comment by sahan29** [comment]
@thisaltennakoon can we get this merged. I'm facing the following error while executing mssql script. This PR is supposed to fix it right?
```
Msg 1785, Level 16, State 0, Line 1736
Introducing FOREIGN KEY constraint 'FK__AM_API_CO__PAREN__090A5324' on table 'AM_API_COMMENTS' may cause cycles or multiple cascade paths. Specify ON DELETE NO ACTION or ON UPDATE NO ACTION, or modify other FOREIGN KEY constraints.
Msg 1750, Level 16, State 1, Line 1736
Could not create constraint or index. See previous errors.
```

**Comment by TheRasanjana** [comment]
The issue is tracked in here https://github.com/wso2/product-apim/issues/10201

**Review by tmkasun — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/Pagination.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package org.wso2.carbon.apimgt.api.model;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/Pagination.java:1]: Add license header.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 2]
```diff
@@ -8000,44 +8002,91 @@ public Comment getComment(ApiTypeWrapper apiTypeWrapper, String commentId, Integ
...
-                    comment.setParentCommentID(resultSet.getString("PARENT_COMMENT_ID"));
-                    comment.setEntryPoint(resultSet.getString("ENTRY_POINT"));
-                    comment.setCategory(resultSet.getString("CATEGORY"));
-                    comment.setReplies(Arrays.asList(getComments(identifier, resultSet.getString("COMMENT_ID"),
-                            connection)));
-                    commentList.add(comment);
+        String sqlQueryForCount;
+        if (parentCommentID == null){
+            sqlQueryForCount  = SQLConstants.GET_ROOT_COMMENTS_COUNT_SQL;
+        }else {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8017]: small formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 3]
```diff
@@ -8000,44 +8002,91 @@ public Comment getComment(ApiTypeWrapper apiTypeWrapper, String commentId, Integ
...
+                        prepStmt.setString(3, identifier.getVersion());
+                        if (parentCommentID != null) {
+                            prepStmt.setString(4, parentCommentID);
+                            // temporary commented below two lines as mssql doesn't support limit and offset
+//                            prepStmt.setInt(5, limit);
+//                            prepStmt.setInt(6, offset);
+                        } else {
+                            // temporary commented below two lines as mssql doesn't support limit and offset
+//                            prepStmt.setInt(4, limit);
+//                            prepStmt.setInt(5, offset);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8049]: shall we create an issue for this and address separately?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 4]
```diff
@@ -8000,44 +8002,91 @@ public Comment getComment(ApiTypeWrapper apiTypeWrapper, String commentId, Integ
...
+                                comment.setReplies(getComments(identifier, resultSet.getString("COMMENT_ID"), 3, 0,
+                                        connection));
+                                list.add(comment);
+                            }
+                                if (limit+offset>total){
+                                    list = list.subList(offset, total);
+                                } else {
+                                    list = list.subList(offset, limit+offset);
+                                }
+
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8072]: Formatting issue?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 5]
```diff
@@ -8000,44 +8002,91 @@ public Comment getComment(ApiTypeWrapper apiTypeWrapper, String commentId, Integ
...
+                                comment.setId(resultSet.getString("COMMENT_ID"));
+                                comment.setText(resultSet.getString("COMMENT_TEXT"));
+                                comment.setUser(resultSet.getString("CREATED_BY"));
+                                comment.setCreatedTime(resultSet.getTimestamp("CREATED_TIME"));
+                                comment.setUpdatedTime(resultSet.getTimestamp("UPDATED_TIME"));
+                                comment.setApiId(resultSet.getString("API_ID"));
+                                comment.setParentCommentID(resultSet.getString("PARENT_COMMENT_ID"));
+                                comment.setEntryPoint(resultSet.getString("ENTRY_POINT"));
+                                comment.setCategory(resultSet.getString("CATEGORY"));
+                                comment.setReplies(getComments(identifier, resultSet.getString("COMMENT_ID"), 3, 0,
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8063]: Shall we move this to a constant? Its better to move this to a configuration. Can address separately.

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8063]: Also I think we don't need to call this when parentCommentID is already set. Shall we filter it using an if block?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 6]
```diff
@@ -8154,27 +8202,48 @@ public boolean editComment(ApiTypeWrapper apiTypeWrapper, String commentId, Comm
...
+    private boolean deleteComment(Identifier identifier, String commentId, Connection connection) throws APIManagementException {
+        int id = -1;
+        String deleteCommentQuery = SQLConstants.DELETE_COMMENT_SQL;
+        String getCommentIDsOfReplies = SQLConstants.GET_IDS_OF_REPLIES_SQL;
+        ResultSet resultSet = null;
+        try {
             id = getAPIID(identifier, connection);
             if (id == -1) {
                 String msg = "Could not load API record for: " + identifier.getName();
                 throw new APIManagementException(msg);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8228]: Let's use an exception code.
```suggestion
                throw new APIManagementException(msg, ExceptionCodes.from(ExceptionCodes.API_NOT_FOUND, existingApiId));
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml [hunk 7]
```diff
@@ -1983,20 +1986,29 @@ paths:
             application/json:
               schema:
                 $ref: '#/components/schemas/Comment'
+        304:
+          description: |
+            Not Modified.
+            Empty body because the client has already the latest version of the requested resource.
+          content: {}
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:1993]: remove if not necessary

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 3 above)
Comment (thisaltennakoon) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: This was solved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 4 above)
Comment (thisaltennakoon) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: This was removed as pagination is now available at the database level

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml (hunk 7 above)
Comment (thisaltennakoon) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/resources/publisher-api.yaml:1993]: The 'Not Modified' response has been added when the user sends an edit request without modifying anything. Should I remove it?

**Pull request merged**